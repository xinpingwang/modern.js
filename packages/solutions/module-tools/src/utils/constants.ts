// generator tsconfig file
export const tempTsconfigName = 'tsconfig.temp.json';
export const cliTsConfigDefaultValue = './tsconfig.json';
