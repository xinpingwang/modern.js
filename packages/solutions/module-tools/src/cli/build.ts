import { Import, Command } from '@modern-js/utils';
import type { PluginAPI } from '@modern-js/core';
import type { IBuildCommandOption } from '../commands/build';

const local: typeof import('../locale') = Import.lazy(
  '../locale/index',
  require,
);
const commands: typeof import('../commands') = Import.lazy(
  '../commands',
  require,
);

export const buildCli = (program: Command, api: PluginAPI) => {
  program
    .command('build')
    .usage('[options]')
    .description(local.i18n.t(local.localeKeys.command.build.describe))
    .option('-w, --watch', local.i18n.t(local.localeKeys.command.build.watch))
    .option(
      '--tsconfig [tsconfig]',
      local.i18n.t(local.localeKeys.command.build.tsconfig),
      './tsconfig.json',
    )
    .option(
      '--style-only',
      local.i18n.t(local.localeKeys.command.build.style_only),
    )
    .option(
      '-p, --platform [platform]',
      local.i18n.t(local.localeKeys.command.build.platform),
    )
    // @deprecated
    // The `--no-tsc` option has been superceded by the `--no-dts` option.
    .option('--no-tsc', local.i18n.t(local.localeKeys.command.build.no_tsc))
    .option('--dts', local.i18n.t(local.localeKeys.command.build.dts))
    .option('--no-clear', local.i18n.t(local.localeKeys.command.build.no_clear))
    .option(
      '-c --config <config>',
      local.i18n.t(local.localeKeys.command.build.config),
    )
    .action(async (subCommand: IBuildCommandOption) => {
      await commands.build(api, subCommand);
    });
};
