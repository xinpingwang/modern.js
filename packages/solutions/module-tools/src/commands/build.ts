import * as path from 'path';
import { fs, Import, dotenv } from '@modern-js/utils';
import type { PluginAPI } from '@modern-js/core';
import onExit from 'signal-exit';
import { tempTsconfigName } from '../utils/constants';
import type { Platform } from '../types';

const tsConfigutils: typeof import('../utils/tsconfig') = Import.lazy(
  '../utils/tsconfig',
  require,
);

const valid: typeof import('../utils/valide') = Import.lazy(
  '../utils/valide',
  require,
);
const buildFeature: typeof import('../features/build') = Import.lazy(
  '../features/build',
  require,
);
/**
 * init work before build task.
 * @param api
 */
export const init = (api: PluginAPI): void => {
  const { appDirectory } = api.useAppContext();

  dotenv.config();

  onExit(() => {
    const tempTsconfigFileAbsPath = path.join(
      appDirectory,
      './node_modules',
      `./${tempTsconfigName}`,
    );
    fs.removeSync(tempTsconfigFileAbsPath);
  });
};

export interface IBuildCommandOption {
  watch?: boolean;
  tsconfig: string;
  platform?: boolean | Exclude<Platform, 'all'>;
  styleOnly?: boolean;
  // @deprecated
  // The `tsc` field has been superceded by the `dts` field.
  tsc: boolean;
  dts?: boolean;
  clear: boolean;
}

export const build = async (
  api: PluginAPI,
  buildCommandOption: IBuildCommandOption,
) => {
  const {
    watch = false,
    tsconfig: tsconfigName,
    tsc,
    dts = false,
    clear,
    platform = false,
    styleOnly = false,
  } = buildCommandOption;

  init(api);

  const { appDirectory } = api.useAppContext();
  const modernConfig = api.useResolvedConfigContext();
  const tsconfigPath = path.join(appDirectory, tsconfigName);
  const outputPath = modernConfig.output.path ?? 'dist';
  const isTsProject = tsConfigutils.existTsConfigFile(tsconfigPath);
  const enableDtsGen = isTsProject && dts;

  valid.valideBeforeTask({ modernConfig, tsconfigPath });

  await buildFeature.build(api, {
    enableWatchMode: watch,
    enableDtsGen,
    isTsProject,
    tsconfigName,
    outputPath,
    styleOnly,
    platform,
    clear,
    legacyTsc: tsc,
  });
};
