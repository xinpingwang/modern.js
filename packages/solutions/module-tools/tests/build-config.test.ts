import { manager } from '@modern-js/core';
import { program } from '@modern-js/utils';
import { buildCli } from '../src/cli/build';
import plugin from '../src';

jest.mock('../src/features/build/normalize', () => {
  const originalModule = jest.requireActual('../src/features/build/normalize');

  return {
    __esModule: true,
    ...originalModule,
    normalizeModuleConfig: (...args: any[]) => {
      const ret = originalModule.normalizeModuleConfig(...args);
      expect(ret).toMatchSnapshot();
      return ret;
    },
  };
});

describe('config in module tools', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });
  it('schema', async () => {
    const main = manager.clone().usePlugin(plugin);
    const runner = await main.init();
    const result = await runner.validateSchema();
    expect(result).toMatchSnapshot();
  });
  it('without build-preset and build-config', async () => {
    const mockAPI = {
      useAppContext: jest.fn((): any => ({
        appDirectory: '',
      })),
      useResolvedConfigContext: jest.fn((): any => ({
        source: {},
        output: {},
        tools: {},
      })),
    };
    const cloned = manager.clone(mockAPI);
    cloned.usePlugin({
      async setup(api) {
        buildCli(program, api);
        program.parse(['', '', 'build']);
      },
    });
    await cloned.init();
  });
});
