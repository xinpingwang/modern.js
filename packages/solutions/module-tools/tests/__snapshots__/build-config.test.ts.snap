// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config in module tools schema 1`] = `
Array [
  Array [
    Object {
      "schema": Object {
        "type": "boolean",
      },
      "target": "source.enableBlockRuntime",
    },
    Object {
      "schema": Object {
        "default": "automatic",
        "enum": Array [
          "classic",
          "automatic",
        ],
      },
      "target": "source.jsxTransformRuntime",
    },
    Object {
      "schema": Object {
        "enum": Array [
          "universal-js",
          "universal-js-lite",
          "browser-js",
          "browser-js-lite",
          "node-js",
        ],
      },
      "target": "output.packageMode",
    },
    Object {
      "schema": Object {
        "typeof": "object",
      },
      "target": "output.packageFields",
    },
    Object {
      "schema": Object {
        "typeof": "boolean",
      },
      "target": "output.enableSourceMap",
    },
    Object {
      "schema": Object {
        "default": "source-code",
        "enum": Array [
          "compiled-code",
          "source-code",
        ],
      },
      "target": "output.importStyle",
    },
    Object {
      "schema": Object {
        "default": "styles",
        "typeof": "string",
      },
      "target": "output.assetsPath",
    },
    Object {
      "schema": Object {
        "else": Object {
          "oneOf": Array [
            Object {
              "properties": Object {
                "bundle": Object {
                  "type": "boolean",
                },
                "bundleOption": Object {
                  "properties": Object {
                    "entry": Object {
                      "type": "object",
                    },
                    "external": Object {
                      "items": Array [
                        Object {
                          "type": "string",
                        },
                      ],
                      "type": "array",
                    },
                    "minify": Object {
                      "enum": Array [
                        "esbuild",
                        "terser",
                        false,
                      ],
                    },
                    "platform": Object {
                      "enum": Array [
                        "node",
                        "browser",
                      ],
                    },
                    "splitting": Object {
                      "type": "boolean",
                    },
                  },
                  "type": "object",
                },
                "bundlessOption": Object {
                  "properties": Object {
                    "sourceDir": Object {
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
                "dts": Object {
                  "type": "boolean",
                },
                "format": Object {
                  "items": Array [
                    Object {
                      "enum": Array [
                        "cjs",
                        "esm",
                        "iife",
                      ],
                    },
                  ],
                  "type": "array",
                },
                "outputPath": Object {
                  "type": "string",
                },
                "target": Object {
                  "enum": Array [
                    "es6",
                    "es5",
                    "es2015",
                    "es2016",
                    "es2017",
                    "es2018",
                    "es2019",
                    "es2020",
                    "esnext",
                  ],
                },
                "tsconfig": Object {
                  "type": "string",
                },
              },
              "type": "object",
            },
          ],
        },
        "if": Object {
          "type": "array",
        },
        "then": Object {
          "items": Array [
            Object {
              "properties": Object {
                "bundle": Object {
                  "type": "boolean",
                },
                "bundleOption": Object {
                  "properties": Object {
                    "entry": Object {
                      "type": "object",
                    },
                    "external": Object {
                      "items": Array [
                        Object {
                          "type": "string",
                        },
                      ],
                      "type": "array",
                    },
                    "minify": Object {
                      "enum": Array [
                        "esbuild",
                        "terser",
                        false,
                      ],
                    },
                    "platform": Object {
                      "enum": Array [
                        "node",
                        "browser",
                      ],
                    },
                    "splitting": Object {
                      "type": "boolean",
                    },
                  },
                  "type": "object",
                },
                "bundlessOption": Object {
                  "properties": Object {
                    "sourceDir": Object {
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
                "dts": Object {
                  "type": "boolean",
                },
                "format": Object {
                  "items": Array [
                    Object {
                      "enum": Array [
                        "cjs",
                        "esm",
                        "iife",
                      ],
                    },
                  ],
                  "type": "array",
                },
                "outputPath": Object {
                  "type": "string",
                },
                "target": Object {
                  "enum": Array [
                    "es6",
                    "es5",
                    "es2015",
                    "es2016",
                    "es2017",
                    "es2018",
                    "es2019",
                    "es2020",
                    "esnext",
                  ],
                },
                "tsconfig": Object {
                  "type": "string",
                },
              },
              "type": "object",
            },
          ],
        },
      },
      "target": "output.buildConfig",
    },
    Object {
      "schema": Object {
        "else": Object {
          "type": "object",
        },
        "if": Object {
          "instanceof": "Function",
        },
      },
      "target": "tools.speedy",
    },
  ],
]
`;

exports[`config in module tools without build-preset and build-config 1`] = `
Array [
  Object {
    "bundle": false,
    "bundlelessOptions": Object {
      "sourceDir": "src",
    },
    "dts": true,
    "format": "esm",
    "ignoreSingleFormatDir": true,
    "outputPath": "./js/treeshaking",
    "outputStylePath": "js/styles",
    "target": "es5",
    "tsconfig": "./tsconfig.json",
    "watch": false,
  },
  Object {
    "bundle": false,
    "bundlelessOptions": Object {
      "sourceDir": "src",
    },
    "dts": true,
    "format": "cjs",
    "ignoreSingleFormatDir": true,
    "outputPath": "./js/node",
    "outputStylePath": "js/styles",
    "target": "es6",
    "tsconfig": "./tsconfig.json",
    "watch": false,
  },
  Object {
    "bundle": false,
    "bundlelessOptions": Object {
      "sourceDir": "src",
    },
    "dts": true,
    "format": "esm",
    "ignoreSingleFormatDir": true,
    "outputPath": "./js/modern",
    "outputStylePath": "js/styles",
    "target": "es6",
    "tsconfig": "./tsconfig.json",
    "watch": false,
  },
]
`;
